### DISCUSSION TITLE
After upgrading to version 1.30.2, the container does not start - unhealthy

### DISCUSSION THREAD

**Discussion by fullcavediver** [dani-garcia/vaultwarden#4295] [Q&A]
Hi,

my Vaultwarden runs behind Trafik 2.24.1 and is updated daily using wathtower. Unfortunately the upgrade didn't work today. Luckily I was able to switch back to version 1.30.1 and everything worked again.

What has changed so massively in the two versions?

I was able to find the following entries in the log file:

```
[2024-01-31 00:08:46.331][vaultwarden::api::web::_][WARN] Parameter guard `p: PathBuf` is forwarding: BadStart('.').
[2024-01-31 00:09:06.304][vaultwarden::api::web::_][WARN] Parameter guard `p: PathBuf` is forwarding: BadStart('.').
[2024-01-31 00:09:37.162][vaultwarden::api::web::_][WARN] Parameter guard `p: PathBuf` is forwarding: BadStart('.').
[2024-01-31 00:10:13.913][request][INFO] GET /admin/.env
[2024-01-31 00:10:13.913][vaultwarden::api::web::_][WARN] Parameter guard `p: PathBuf` is forwarding: BadStart('.').
[2024-01-31 00:10:13.914][response][INFO] (web_files) GET /<p..> [10] => 422 Unprocessable Entity
[2024-01-31 00:10:38.698][vaultwarden::api::web::_][WARN] Parameter guard `p: PathBuf` is forwarding: BadStart('.').
[2024-01-31 04:11:01.664][rocket::server][WARN] Received SIGTERM. Requesting shutdown.
[2024-01-31 04:11:01.669][vaultwarden][INFO] Vaultwarden process exited!
[2024-01-31 04:11:05.442][start][INFO] Rocket has launched from http://0.0.0.0:80
```

The compose file is also very simple:

```
version: '3'

services:
  vaultwarden:
    image: vaultwarden/server:1.30.1
    container_name: vaultwarden
    hostname: vaultwarden
    restart: unless-stopped
    volumes:
      - /vaultwarden:/data
      - /vaultwarden/.env:/.env
    environment:
      - LOG_FILE=/data/bitwarden.log
      - LOG_LEVEL=info
      - EXTENDED_LOGGING=true
    labels:
      - "traefik.enable=true"
      - "traefik.http.middlewares.vaultwarden-https.redirectscheme.scheme=https"
      - "traefik.http.routers.vaultwarden-http.entrypoints=web"
      - "traefik.http.routers.vaultwarden-http.rule=Host(`domain-1`) || Host(`domain-2`)"
      - "traefik.http.routers.vaultwarden-http.middlewares=vaultwarden-https@docker"
      - "traefik.http.routers.vaultwarden.entrypoints=web-secure"
      - "traefik.http.routers.vaultwarden.rule=Host(`domain-1`) || Host(`domain-2`)"
      - "traefik.http.routers.vaultwarden.tls=true"
      - "traefik.http.routers.vaultwarden.tls.certresolver=default"
      - "traefik.http.routers.vaultwarden.middlewares=secHeaders@file"
      - "traefik.http.services.vaultwarden.loadbalancer.server.port=80"
    networks:
      - traefik_proxy
      - default
networks:
  traefik_proxy:
    name: traefik_proxy
    external: true
  default:
    driver: bridge 
```

Please help me find the error. Thank you very much!

**Comment by BlackDex** [comment]
Something externally is stopping the container.
Try to run the compose manually without the `-d`, so `docker compose up`

For a full example first fully stop all, and start it again:
```bash
docker compose down

docker compose up
```

**Comment by fullcavediver** [comment]
unfortunately no further information in the log file...

```
docker-compose -f ./vaultwarden.yml up
[+] Running 2/2
 ✔ Network vaultwarden_default  Created                                                                                                                                                                                                 0.1s
 ✔ Container vaultwarden        Created                                                                                                                                                                                                 0.1s
Attaching to vaultwarden
vaultwarden  | /--------------------------------------------------------------------\
vaultwarden  | |                        Starting Vaultwarden                        |
vaultwarden  | |                           Version 1.30.2                           |
vaultwarden  | |--------------------------------------------------------------------|
vaultwarden  | | This is an *unofficial* Bitwarden implementation, DO NOT use the   |
vaultwarden  | | official channels to report bugs/features, regardless of client.   |
vaultwarden  | | Send usage/configuration questions or feature requests to:         |
vaultwarden  | |   https://github.com/dani-garcia/vaultwarden/discussions or        |
vaultwarden  | |   https://vaultwarden.discourse.group/                             |
vaultwarden  | | Report suspected bugs/issues in the software itself at:            |
vaultwarden  | |   https://github.com/dani-garcia/vaultwarden/issues/new            |
vaultwarden  | \--------------------------------------------------------------------/
vaultwarden  |
vaultwarden  | [INFO] Using environment file `.env` for configuration.
vaultwarden  |
vaultwarden  | [INFO] Using saved config from `data/config.json` for configuration.
vaultwarden  |
vaultwarden  | [2024-01-31 12:12:41.345][start][INFO] Rocket has launched from http://0.0.0.0:80
```

browser shows: `404 page not found`

**Reply by BlackDex** [reply]
Looks like your reverse proxy isn't reaching Vaultwarden. Check it's config

**Comment by fullcavediver** [comment]
But nothing has changed in the config of traefik, with version 1.30.1 it works without any problems - I use the same config for all public services. Does vaultwarden need special setings from version 1.30.2?

I have attached my docker compose file above - what is wrong there?

**Reply by BlackDex** [reply]
Well, your traefik is probably also updated. Maybe something there has changed?

Nothing changed in regards to connecting to Vaultwarden.
I use Nginx, and changed nothing and it's still running fine.

**Reply by BlackDex** [reply]
Also, of Vaultwarden does not show logs of connects, then either the log level is to low, or nothing is reaching Vaultwarden, and must be something on the traefik side.

Check the logs of both during a request.

**Comment by tjmv** [comment]
I have the same problem and I agree that something must have changed in today's update to cause this. However, I should also mention that when accessing Vaultwarden directly (i.e. making the ports accessible), it works fine. As far as I know Traefik only picks up hosts that are healthy, but the Vaultwarden container never reaches that state. I can only speculate here as I haven't looked into the issue and simply decided to downgrade to avoid a longer downtime, but my guess is that there is a healthcheck being done when the service is not yet available via the reverse proxy that also relies on the health of the container.

**Reply by BlackDex** [reply]
Increase the log level to debug, and see if that provides any answers. It could very well be something in your config for some reason. But without logs, we can't help.

**Reply by tjmv** [reply]
That's the thing. The container does not report that it's healthy so it does not get picked up by Traefik. The last thing in the log is simply that Rocket has launched, but that was already posted above. The only difference with 1.30.1 is that it gets a request to make it healthy that the latest version does not get. This is from the previous version, the last two lines are not present in 1.30.2:
```
vaultwarden  | [2024-01-31 16:58:44.955][start][INFO] Rocket has launched from http://0.0.0.0:8000
vaultwarden  | [2024-01-31 16:59:44.954][vaultwarden::api::core::two_factor][DEBUG] Sending notifications for incomplete 2FA logins
vaultwarden  | [2024-01-31 16:59:44.955][vaultwarden::api::core::accounts][DEBUG] Purging auth requests
vaultwarden  | [2024-01-31 16:59:45.010][request][INFO] GET /alive
vaultwarden  | [2024-01-31 16:59:45.010][response][INFO] (alive) GET /alive => 200 OK
```

**Reply by tjmv** [reply]
See @owine's comment below. That's what you get if you manually run the healthcheck script. The result is that the container never reports as healthy and is never picked up by Traefik (and other services using a similar method).

**Comment by owine** [comment]
Traefik waits for the container to report healthy when it has a health check. I am seeing the container constantly report unhealthy with the following error
```
/healthcheck.sh: 17: .: .env: not found
```
But the file is present at `/.env`

**Reply by stefan0xC** [reply]
Seems to be caused by https://superuser.com/questions/1628109/sh-source-cant-find-files so we should either default to `/.env` or use `#!/bin/bash` in the `healthcheck.sh`

**Reply by owine** [reply]
I've confirmed that either of these changes restores the healthcheck to a working state and restores my VW to load properly. Thanks for suggesting

**Comment by fullcavediver** [comment]
thank you @tjmv for your feedback.

> Well, your traefik is probably also updated. Maybe something there has changed?

I think I can rule that out. Last traefik update: 2024-01-27 04:56:16 build: Docker 20.10.23 on linux, amd64

Apart from that, all other containers work and vaultwarden with the image 1.30.1 also works. I don't see any error messages in the Traefik log file either.

**Reply by BlackDex** [reply]
Ok, thx, always good to check and not assume it isn't an issue.

**Comment by fullcavediver** [comment]
> so we should either default to /.env or use #!/bin/bash in the healthcheck.sh

in the vaultwarden container?

Will this happen again with the next automatic update with watchtower?

**Reply by saltydk** [reply]
He means that it was broken in https://github.com/dani-garcia/vaultwarden/pull/4143/files so they should fix it in the image.

**Reply by saltydk** [reply]
You can probably fix it by setting the ENV ENV_FILE=/.env

**Reply by owine** [reply]
This works as well and is probably the easiest workaround to drop into a compose instead of mounting in an edited healthcheck script until the script shipped with the image is fixed up.

**Comment by fullcavediver** [comment]
thank you @saltydk

> You can probably fix it by setting the ENV ENV_FILE=/.env

you mean in the docker image file? I'm overwhelmed with that :-) I'd rather leave it to the professionals and I'm waiting for image update or 1.30.3 :-D

**Reply by BlackDex** [reply]
No, you can set that as en env variable just like all other env variables.

Just in your docker-compose.yaml.

**Comment by BlackDex** [comment]
Hmm, strange, i have tested it with and without setting that env though, and having the .env file in multiple locations.

Which base image are you using?

Also, switching to bash is not an option i think, rather keep it Posix compatible. Also because I'm still planning on having a Busybox or Google Distroless addition.

**Reply by owine** [reply]
`vaultwarden/server:latest` image in use.

**Comment by fullcavediver** [comment]
Cool! Thanks, friends ! You are the best!

I only had volume mapping:

```
volumes:
      - /vaultwarden/.env:/.env
```

but no environment variable:

```
environment:
      - ENV_FILE=/.env
```

Strange... the .env file was there and was read when the container started...

```
vaultwarden  | [INFO] Using environment file `.env` for configuration.
vaultwarden  |
vaultwarden  | [INFO] Using saved config from `data/config.json` for configuration.
```

> Which base image are you using?

Until today I have always used the latest image file, after the problems I changed to vaultwarden/server:1.30.1 and now back to vaultwarden/server:latest

**Reply by BlackDex** [reply]
Well, the healthcheck didn't used that file previously. That is what changed. Maybe we should change the shebang to `#!/usr/bin/env sh` instead.

Ill check what happens and provide a PR to fix this.

**Comment by fullcavediver** [comment]
In my setup the file is in the root directory of VW data:

```
    volumes:
      - ./vaultwarden:/data
      - ./vaultwarden/.env:/.env
    environment:
      - LOG_FILE=/data/bitwarden.log
      - LOG_LEVEL=info
      - EXTENDED_LOGGING=true
      - ENV_FILE=/.env
```

After I added the environment variable, the status of the container changed to 'healthy' and the web interface worked again

**Comment by BlackDex** [comment]
Also, the strange thing is, the script does see that the file exits, else it wouldn't even try to load it.
And it does seem to work on Alpine, but breaks on Debian.

**Reply by saltydk** [reply]
Is it strange? Isn't sh a symlink to ash in alpine?

**Reply by saltydk** [reply]
Alpine:
```
stat /bin/sh
File: '/bin/sh' -> '/bin/busybox'
Size: 12              Blocks: 0          IO Block: 4096   symbolic link
Device: 100086h/1048710d        Inode: 5661142     Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-01-26 18:53:03.000000000 +0100
Modify: 2024-01-26 18:53:03.000000000 +0100
Change: 2024-01-31 22:09:54.776797242 +0100
```

Debian:
```
stat /usr/bin/sh
File: /usr/bin/sh -> dash
Size: 4               Blocks: 0          IO Block: 4096   symbolic link
Device: 0,390   Inode: 5653090     Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-01-05 14:20:48.000000000 +0100
Modify: 2023-01-05 14:20:48.000000000 +0100
Change: 2024-01-31 22:08:31.799133100 +0100
Birth: 2024-01-31 22:08:31.799133100 +0100
```

**Reply by BlackDex** [reply]
What does that has to do with `if [ -r .env ]` being valid, but then `. .env` breaking and not seeing the file?
I would expect the if to fail in those cases.

It doesn't matter which shell is used, that is why the script is POSIX valid.

**Reply by stefan0xC** [reply]
There seems to be a subtle difference to the path lookup of the [`.`](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_18) command which according to the POSIX standard depends on the `$PATH` (if there is no `/` in the given argument), i.e. in one shell the current folder seems to be implicitly added in the other you have to explicitly `export PATH=$PATH:` for `. .env` to work. Meanwhile the path lookup of [`test`](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/test.html#tag_20_128) does not care about the `$PATH` variable.

**Reply by saltydk** [reply]
Pretty sure sourcing .env is not POSIX compliant if that file is not in PATH where as /.env is the compliant way. Ash does not follow POSIX to a tee.

**Reply by BlackDex** [reply]
Thanks for the explanation! Learned something about the whole nice POSIX not-so-much-compatible-with-all standard :p

**Reply by owine** [reply]
Your fix in the `:testing` image is working for me. Thanks!
